### PR TITLE
Added null- and exception-safe variants of (priority)readOnly and modify

### PR DESCRIPTION
**Pull request by tivervac**
Also added I[Read|Write]Access where applicable

This is based on the feedback of #677

### CODE REVIEW

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/AbstractReadWriteAcces.java [hunk 1]
```diff
@@ -17,11 +17,11 @@
  * @author Sven Efftinge - Initial contribution and API
  * @author Jan Koehnlein
  */
-public abstract class AbstractReadWriteAcces<P> implements IReadAccess<P> {
+public abstract class AbstractReadWriteAcces<State> implements IReadAccess<State>, IWriteAccess<State> {
```
Comment (szarnekow) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/AbstractReadWriteAcces.java]: Thank you for renaming the type parameters. I like the attention to detail 👍

File: org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/IReadAccess.java [hunk 2]
```diff
@@ -8,22 +8,140 @@
...
+	 * @param exceptionHandler Supplies a result in case an exception is raised during execution
+	 * 
+	 * @return The result of executing {@code work},
+	 *         the result of querying {@code defaultResult} if the State is null, or
+	 *         the result of executing {@code exceptionHandler} in case an exception is raised
+	 * @since 2.14
+	 */
+	default <Result> Result tryReadOnly(
+		IUnitOfWork<Result, State> work,
+		Supplier<Result> defaultResult,
```
Comment (szarnekow) [inline org.eclipse.xtext.util/src/org/eclipse/xtext/util/concurrent/IReadAccess.java]: We could use Supplier<? extends Result> similar to the signature of the exceptionHandler.

Same for the other try's

**Review by szarnekow — APPROVED** [review]
LGTM

**Comment by cdietrich** [comment]
@szarnekow what is the state of this pr?

**Comment by szarnekow** [comment]
Ready to be merged.

**Comment by cdietrich** [comment]
thx @tivervac

**Comment by cdietrich** [comment]
hi can you please add the sample project?

**Comment by szarnekow** [comment]
I assume that the `@Delegate` does not implement a correct detection of equivalent signatures, e.g type parameter-names are not anonymised.

```
interface J<P, Q> {
}

interface I<X> {
	def <Y> Y m(J<Y, X> j);
}

class C<T> implements I<T> {
	@Delegate
	I<T> delegate;
	
	override <V> V m(J<V, T> j) {}
} 
```

Reproduces this bug.

**Comment by cdietrich** [comment]
i am not sure reverting the change would be a good idea....

**Comment by szarnekow** [comment]
Neither am I. We should rather file a ticket for the broken impl in `@Delegate` and try to fix that.

**Comment by cdietrich** [comment]
created https://github.com/eclipse/xtext/issues/2246

**Comment by sailingKieler** [comment]
> Neither am I. We should rather file a ticket for the broken impl in @Delegate and try to fix that.

I would prefer that, too.

**Pull request merged**